pkg/services/store/entity/sqlstash/sql_storage_server.go

```diff
@@ -1106,12 +1166,325 @@ func (s *sqlEntityServer) List(ctx context.Context, r *entity.EntityListRequest)
+		err = w.Send(batch)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// watch is a helper to get the next set of entities and send them to the client
+func (s *sqlEntityServer) watch(ctx context.Context, r *entity.EntityWatchRequest, w entity.EntityStore_WatchServer) error {
+	s.log.Debug("watch", "since", r.Since)
+
+	rr := &entity.ReadEntityRequest{
+		WithBody:   r.WithBody,
+		WithStatus: r.WithStatus,
+	}
+
+	fields := s.getReadFields(rr)
+
+	entityQuery := selectQuery{
+		dialect: s.dialect,
+		fields:  fields,
+		from:    "entity_history", // the table
+		args:    []any{},
+		limit:   100, // r.Limit,
+		// offset:   0,
+		oneExtra: true, // request one more than the limit (and show next token if it exists)
+	}
+
+	// TODO fix this
+	// entityQuery.addWhere("namespace", user.OrgID)
+
+	/*
+		if len(r.Group) > 0 {
+			entityQuery.addWhereIn("group", r.Group)
+		}
+	*/
+
+	if r.Since > 0 {
+		entityQuery.addWhere("resource_version > ?", r.Since)
+	}
+
+	if len(r.Resource) > 0 {
+		entityQuery.addWhereIn("resource", r.Resource)
+	}
+
+	if len(r.Key) > 0 {
+		where := []string{}
+		args := []any{}
+		for _, k := range r.Key {
+			key, err := entity.ParseKey(k)
+			if err != nil {
+				return err
+			}
+
+			args = append(args, key.Namespace, key.Group, key.Resource)
+			whereclause := "(" + s.dialect.Quote("namespace") + "=? AND " + s.dialect.Quote("group") + "=? AND " + s.dialect.Quote("resource") + "=?"
+			if key.Name != "" {
+				args = append(args, key.Name)
+				whereclause += " AND " + s.dialect.Quote("name") + "=?"
+			}
+			whereclause += ")"
+
+			where = append(where, whereclause)
+		}
+
+		entityQuery.addWhere("("+strings.Join(where, " OR ")+")", args...)
+	}
+
+	// Folder guid
+	if r.Folder != "" {
+		entityQuery.addWhere("folder", r.Folder)
+	}
+
+	// if we have a page token, use that to specify the first record
+	/*
+		continueToken, err := GetContinueToken(r)
+		if err != nil {
+			return err
+		}
+		if continueToken != nil {
+			entityQuery.offset = continueToken.StartOffset
+		}
+	*/
+
+	if len(r.Labels) > 0 {
+		var args []any
+		var conditions []string
+		for labelKey, labelValue := range r.Labels {
+			args = append(args, labelKey)
+			args = append(args, labelValue)
+			conditions = append(conditions, "(label = ? AND value = ?)")
+		}
+		query := "SELECT guid FROM entity_labels" +
+			" WHERE (" + strings.Join(conditions, " OR ") + ")" +
+			" GROUP BY guid" +
+			" HAVING COUNT(label) = ?"
+		args = append(args, len(r.Labels))
+
+		entityQuery.addWhereInSubquery("guid", query, args)
+	}
+
+	entityQuery.addOrderBy("resource_version", Ascending)
+
+	query, args := entityQuery.toQuery()
+
+	rows, err := s.sess.Query(ctx, query, args...)
+	if err != nil {
+		return err
+	}
+	defer func() { _ = rows.Close() }()
+
+	batch := &entity.EntityWatchResponse{
+		Action:    entity.Entity_UNKNOWN,
+		Entity:    []*entity.Entity{},
+		Timestamp: time.Now().UnixMilli(),
+	}
+
+	for rows.Next() {
```

Comment:
Just double-checking, this is supposed to have the `for hasmore := true; hasmore;` logic as well, right?

Reply:
added hasmore logic